src/commonMain/kotlin/fr/acinq/lightning/payment/Bolt12Invoice.kt

```diff
@@ -125,7 +126,33 @@ data class Bolt12Invoice(val records: TlvStream<InvoiceTlv>) : PaymentRequest()
             return Bolt12Invoice(TlvStream(tlvs + Signature(signature), request.records.unknown + customTlvs))
         }
 
-        fun validate(records: TlvStream<InvoiceTlv>): Either<InvalidTlvPayload, Bolt12Invoice> {
+        sealed class Bolt12ParsingResult {
+            data class Success(val invoice: Bolt12Invoice) : Bolt12ParsingResult()
+            sealed class Failure : Bolt12ParsingResult() {
+                data class Malformed(val invalidTlvPayload: InvalidTlvPayload) : Failure()
+                data class RecipientError(val invoiceError: InvoiceError) : Failure()
+            }
+        }
+
+        fun extract(records: TlvStream<OnionMessagePayloadTlv>): Bolt12ParsingResult {
```

Comment:
Converting a list of `OnionMessagePayloadTlv` to a `Bolt12Invoice` does not belong to the `OfferManager`, it's generic parsing code. I hesitated between moving it to `Bolt12Invoice` or `OnionMessage`.